File: linked-list/01-initial-concept.ts
// ==============================================================================================

let obj1: { [key: string]: string | number } = {
    name: 'ahmed',
};
let data = obj1;
data.age = 22;
console.log(obj1);

// OUTPUT
// { name: 'ahmed', age: 22 }

// ==============================================================================================

let obj: { [key: string]: string | number } = {
    name: 'ahmed',
    age: 22,
}
let arr1 = [1, 2, 3, 4]
let myName = "AHMED FARAZ"

/**
 * @property {Function} passByReference to pass variables 
 * to check which follows pass-by-reference property 
 * @param {{ [key: string]: string | number }} objt 
 * @param {number[]} arrr 
 * @param {string} myName 
 * @returns {void}
 */
const passByReference: (
    objt: { [key: string]: string | number },
    arrr: number[],
    myName: string
) => void = (objt, arrr, myName) => {
    objt.profession = 'cloud expert';
    arrr.push(5);
    arrr.unshift(0);
    myName = "STEVE JOBS";
    console.log('INSIDE FUNCTION');
    console.log(obj, arr1, myName);
}

console.log('BEFORE FUNCTION');
console.log(obj, arr1, myName);
passByReference(obj, arr1, myName);
console.log('AFTER FUNCTION');
console.log(obj, arr1, myName);

// OUTPUT
// BEFORE FUNCTION
// { name: 'ahmed', age: 22 } [ 1, 2, 3, 4 ] AHMED FARAZ
// INSIDE FUNCTION
// { name: 'ahmed', age: 22, profession: 'cloud expert' } [ 0, 1, 2, 3, 4, 5 ] STEVE JOBS 
// AFTER FUNCTION
// { name: 'ahmed', age: 22, profession: 'cloud expert' } [ 0, 1, 2, 3, 4, 5 ] AHMED FARAZ

// ==============================================================================================
import { LinkedList, LinkedListNode } from './utils/types';

let arr2 = [0, 1, 2, 3, 4, 5]; // converting this array into linked list
let myLinkedList = new LinkedList();

/**
 * @property {Function} insertNode function to insert a node at the end of the link list 
 * and return back the head
 * @param {LinkListNode} head pass by reference
 * @param {number} data data to be inserted at the targeted location (here in the end)
 * @returns {LinkedListNode} pointer to the head node
 */
const insertNode: (
    head: LinkedListNode | null,
    data: number
) => LinkedListNode = (head, data) => {
    console.log('DATA >> ', JSON.stringify(head), data)
    if (head == null) {
        head = new LinkedListNode(data);
        return head;
    }
    let currentHead = head;
    while (currentHead.next) {
        currentHead = currentHead.next
    }
    currentHead.next = new LinkedListNode(data);
    return head;
}

arr2.forEach(num => {
    myLinkedList.head = insertNode(myLinkedList.head, num);
    console.log("CURR HEAD >> ", JSON.stringify(myLinkedList.head), '\n');
})
console.log("FINAL LLIST >> ", JSON.stringify(myLinkedList));

// OUTPUT
// DATA >>  null 0
// CURR HEAD >>  {"data":0,"next":null}

// DATA >>  {"data":0,"next":null} 1
// CURR HEAD >>  {"data":0,"next":{"data":1,"next":null}}

// DATA >>  {"data":0,"next":{"data":1,"next":null}} 2
// CURR HEAD >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":null}}}

// DATA >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":null}}} 3
// CURR HEAD >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":{"data":3,"next":null}}}}

// DATA >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":{"data":3,"next":null}}}} 4
// CURR HEAD >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":{"data":3,"next":{"data":4,"next":null}}}}}

// DATA >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":{"data":3,"next":{"data":4,"next":null}}}}} 5
// CURR HEAD >>  {"data":0,"next":{"data":1,"next":{"data":2,"next":{"data":3,"next":{"data":4,"next":{"data":5,"next":null}}}}}}

// FINAL LLIST >>  {"head":{"data":0,"next":{"data":1,"next":{"data":2,"next":{"data":3,"next":{"data":4,"next":{"data":5,"next":null}}}}}}}

File: linked-list/01-initial-concept.py
# ==============================================================================================
person1 = {
    "name": "Ahmed Faraz",
}
person2 = person1
person2["age"] = 22
print(person1)

# OUTPUT
# {'name': 'Ahmed Faraz', 'age': 22}

# ==============================================================================================
my_string: str = "ahmed faraz"
my_obj: object = {
    "country": "Pakistan"
}
my_list: list = ["karachi", "lahore", "islamabad"]

# Function to pass variables to check which follows pass-by-reference property


def pass_by_ref(str: str, obj: object, lst: list):
    str = "steve jobs"
    obj["country"] = "Australia"
    lst.append("Sidney")
    lst.append("Melbourne")
    print('INSIDE FUNCTION')
    print(my_string, my_obj, my_list)


print('BEFORE FUNCTION')
print(my_string, my_obj, my_list)
pass_by_ref(my_string, my_obj, my_list)
print('AFTER FUNCTION')
print(my_string, my_obj, my_list)

# OUTPUT
# BEFORE FUNCTION
# ahmed faraz {'country': 'Pakistan'} ['karachi', 'lahore', 'islamabad']
# INSIDE FUNCTION
# ahmed faraz {'country': 'Australia'} ['karachi', 'lahore', 'islamabad', 'Sidney', 'Melbourne']
# AFTER FUNCTION
# ahmed faraz {'country': 'Australia'} ['karachi', 'lahore', 'islamabad', 'Sidney', 'Melbourne']

# ==============================================================================================

# class: Linked List


class LinkedList:
    def __init__(self):
        self.head = None

    def print_list(self):
        print('The Linked List', end=" ")
        curr_node = self.head
        while(curr_node):
            print("->", curr_node.data, end="")
            curr_node = curr_node.next


class Node:  # class: Node
    def __init__(self, data):
        self.data = data
        self.next = None


# Function:to insert node at the end of a linked list
def insert_node(head: None | Node, data: int) -> None | Node:
    if(head == None):
        head = Node(data)
        return head
    curr_node = head
    while(curr_node.next):
        curr_node = curr_node.next
    curr_node.next = Node(data)
    return head


# create a linked list
my_llist = LinkedList()

# insert elements
num_list = [1, 2, 3, 4, 5]
for num in num_list:
    my_llist.head = insert_node(my_llist.head, num)

# print the list
my_llist.print_list()

# OUTPUT
# The Linked List -> 0-> 1-> 2-> 3-> 4
